### PR TITLE
Created back-end for the User event many-to-many tables

### PR DESCRIPTION
**Pull request by FrederickMih**
- This Branch consists of building a private event project in which a user can sign up and create an event.
- Other users can join these events and see the upcoming and past events.
- We added the signup, login, and logout links
- We used sessions to store the current user's session.
- We used helper methods to improve the performance, by putting the views logic in the helpers.
- We used Rspec to test the models of our database

image
image
image
image
image
image
image
image
image

### CODE REVIEW

File: app/views/layouts/_header.html.erb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    <ul>
+      <li><%= link_to "Log out", remsession_event_path(1), method: :delete %></li>
+      <li><%= link_to "Home", user_path(current_user.id) %></li>
+    </ul>
+    <% else %>
+    <ul>
+        <li><%= link_to "Log in", login_path %></li> 
+        <li><%= link_to "Sign up", sign_up_path %></li>
+      </ul>
+      <% end %>
```
Comment (acushlakoncept) [inline app/views/layouts/_header.html.erb:21]: There should be absolutely NO logic in views.

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello @FrederickMih and @AymanJabr, both of you have done an awesome job 👍 

* Your PR title and descriptions are super active. 👏 
* Well detailed ReadMe file 👏 

Let's make a couple of changes, and we will be done.

#### Behaviour

- [x] There is an issue signing up in your application. 
Screen Shot 2021-04-02 at 8 19 07 PM. I noticed the method didn't need any argument but you gave one to the argument call, removing it should fix it. output_user_errors _~(@user)~_

- [x] The All Events page, as the name implies should list all events created, categorizing them into past and upcoming events instead of all users. Event's index page displays past and upcoming events lists ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

- [x] The event's show page displays the list of attendants ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/events_attendees_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

- [x] Looking at the above request means there should be a button for a user to be able attend an event in other to make the attendants list.

- [x] The user's show page displays events the user has created and events that the user is attending to separated in past and upcoming ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

#### Associations

- [x] Indexes are built to optimize the access of your SQL queries, in your `event_user` table, since the application will involving querying both fields, kindly index both.. example 
```
    t.belongs_to :user, index: true
    t.belongs_to :event, index: true
```

#### Controllers

- [x] Rubocop helps enforce writing good Ruby code, inside `application_controller` kindly enable back Rubocop by removing `# rubocop:disable Style/GuardClause` and refactoring your code with the given suggestion "Use a guard clause (return if logged_in?) instead of wrapping the code inside a conditional expression."

##### EventsController
###### `index` action
Take advantage of the scope(`past and upcoming`) you created in your event model and pass the variable directly to your view calling it as below
- [x] lists past events, i.e. `@past_events = Event.past` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [x] lists ucomping events, i.e. `@upcoming_events = Event.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

##### UsersController
###### `show` action
Here you can also take advantage of the scope(`past and upcoming`) you created in your event model and pass the variable directly to your view calling it as below

- [x] lists events created by user, i.e. `@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [x] lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- [x] lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

##### Views
- [x] There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it.  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from Pathwright)
I have gone ahead and made some comments on your codebase see affected files below:

**Don't know where to start or how to?** See  [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4) 

- [x] This is a lot of disabling in your `event_helper` file
```
# rubocop:disable Metrics/ModuleLength
# rubocop:disable Lint/UselessAssignment
# rubocop:disable Style/GuardClause
# rubocop:disable Lint/Void
```
Kindly see how you can refactor your code and reduce the number of cops your your disabling.

#### Tests
I see you have nicely implemented unit test for users and events spec, you should try [shoulda-matchers](https://github.com/thoughtbot/shoulda-matchers), it has a more expressive language for unit testing, it's a perfect gem for it. See example code below

```ruby
# RSpec
RSpec.describe MenuItem, type: :model do
  describe 'associations' do
    it { should belong_to(:category).class_name('MenuCategory') }
  end

  describe 'validations' do
    it { should validate_presence_of(:name) }
    it { should validate_uniqueness_of(:name).scoped_to(:category_id) }
  end
end
```

- [ ] authentication should have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))
- [ ] event management features should have integration tests (Rspec + Capybara) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/integration_tests_requirement.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49723079/))


##### Nice to have [OPTIONAL]

- [x] Add a little CSS styles to your app in order to give it a little face lift. 

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla) or reply to this thread, if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

File: app/views/layouts/_header.html.erb (hunk 1 above)
Comment (AymanJabr) [reply app/views/layouts/_header.html.erb:21]: Most of what you requested has already been implemented, are you sure you are looking at the right branch? our branch is called events-private, and not private-events

**Review by oloomoses — CHANGES_REQUESTED** [review]
# Changes Required :recycle: 
**Hello team, great job working on some of the raised issues from the previous review :muscle:** 
**Some changes still required before the project can be approved**

- [x] Kindly add a link to the event's show page for each event. The event's show page displays the list of attendants
- [x] Kindly include a link to the event's index page in the navigation. Event's index page displays past and upcoming events lists
- [x] Still getting this :point_down:  error when trying to signup. Kindly fix this
Screenshot from 2021-04-03 14-06-07

Kindly implement the above changes before submitting another review.
**Feel free to reach me on slack or leave a comment while mentioning me and I will be happy to reply**

_Slack: Oloo Moses_

**Comment by AymanJabr** [comment]
The way we designed the website, is that each user can see the events that he is going to. And he can gain access to an event, only if the creator of the event has sent him a link (In this case the id of the event that he made). So the event Index page shows only those events that a user has gone to previously, or is planning to go to.

**Review by smcommits — CHANGES_REQUESTED** [review]
# Status: Changes Requested 
## Hello, @FrederickMih and @AymanJabr! You guys have done a great job. The project is almost complete. To make things even better, we still need to address some issues that were pointed out by the previous TSEs. 

## Required Changes
 
- [x] **Users should be able to access the event's show page and it should display all the attendees**

Currently, upon visiting the event's show page, I get the following errors
error

I really appreciate your efforts to add the feature that you have mentioned in the previous PR message. Great work. But, unfortunately, as far as I understood your feature, currently I see no option for a user to invite other users and for other users to attend an event and see other attendees. This is a mandatory project requirement. Please implement an event's show page with a list of attendees and some call to action to attend the event. 

It is very possible that I have overlooked something and failed to understand the feature you have mentioned in the PR message. I request you to please describe it elaborately and if possible please demonstrate how can a user invite other users, how can a user attend the events, and lastly, how can a user see the event's show page and see the list of all attendees. 

# 
Please reach out to me on Slack if you need any assistance.

**Review by rindrajosia — CHANGES_REQUESTED** [review]
Changes Required :red_circle:
To Highlight: :angel:

- Good job, User can attend upcoming events now :+1: 

To Improve :factory_worker: (REQUIRED)

- [x] User as event creator: 

1. User model: has_many :events, foreign_key: :creator_id ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
2. Event model belongs_to :creator, class_name: "User" ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
(Please make sure to follow the project specifications about the FKs. In your Event and User model)

- Following the previous point: Please keep in mind that after you modify the Event model you'll have to modify your migrations so that it reflects correctly the foreign keys and indexes in your schema

1. events table has **creator_id** as the foreign key ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
2. the events table has column creator_id of type integer
3. and an index for creator_id column should be added

- [x] EventsController#index action:
- lists past events, i.e. @past_events = Event.previous_events ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
-  lists ucomping events, i.e. @upcoming_events = Event.upcoming_events ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

- [ ] UsersController#show action:

- lists events created by user, i.e. @created_events = current_user.events ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
- lists events previoulsy attendented by user **using event model scope**, i.e. @past_events = current_user.attended_events.previous ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
-  lists upcoming events previously attended by user **using event model scope**, i.e. @upcoming_events = current_user.attended_events.upcoming ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

- [ ] As The user's show page should display events that the user was attending in past  ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)), please make sure that user can also attend past event.
- [x] Some of your unit tests don't pass, please fix it because, it means that there are some bug in your app:


That's it, you can submit another code review request after making these changes :recycle: :innocent:
With all my respect :pray:
Rindra Josia

**Comment by AymanJabr** [comment]
image
1) The project is already conforming to the requirements, we are already making use of foreign keys, class names, and and sources.
2) A has_many relationship does not have foreign keys, it holds them in the linking table.
image
We already added past and upcoming events

image
This makes no sense, in the context of the project or otherwise.

image
image
image
We are already making use of the association `user.events` in order to show all events created by a user. Same for `user.attended_events.past` and `user.attended_events.upcoming`.

File: app/helpers/event_helper.rb [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    user.attended_events.past.each do |event|
+      content = content_tag(:tr, '')
+      content << content_tag(:td, event.name)
+      content << content_tag(:td, event.description)
+      content << content_tag(:td, event.date)
+      content << content_tag(:br)
+      all_contents << content
+    end
+    all_contents
+  end
```
Comment (smcommits) [inline app/helpers/event_helper.rb:120]: You are repeating yourself and creating some content in each method. You can easily refactor the creation of content in a method. Let that method take dynamic input and you can pass in a hash with values. This way you won't be repeating yourself and you will only have to worry about one place while maintaining this particular part of your application.

**Review by smcommits — APPROVED** [review]
# Status: Approved :+1: 
## Hello, team! Nice work implementing the changes requested by the previous TSEs. Your application appears to be amazing. This PR is now approved. You can go ahead and merge the branch. Congratulations and good luck with your future projects. 

## Optional 
#### Please note that the following suggestions are optional. 

- [ ] **Your code is not very DRY in some places**

Please refer to the inline comments below

**Pull request merged**